components/_impl/workers/subprocess_rpc.py

```diff
@@ -0,0 +1,488 @@
+
+    Instead of trying to interrupt the pipe read, we can cause it terminate by
+    writing to the pipe; because we control the read (via `Pipe.read`) we can
+    catch the sentinel timeout value and raise appropriately.
+
+    This class is designed to be extremely lightweight. Timeouts should be on
+    the order of seconds (or minutes), and are only to prevent deadlocks in the
+    case of catastrophic worker failure. As a result, we prioritize low
+    resource usage over the ability to support small timeouts.
+    """
+
+    _singleton_lock = threading.Lock()
+    _singleton: typing.Optional["_TimeoutPIPE"] = None
+
+    _loop_lock = threading.Lock()
+    _active_reads: typing.Dict[int, typing.Tuple[float, float]]
+    _loop_cadence = 1  # second
+
+    @classmethod
+    def singleton(cls) -> "_TimeoutPIPE":
+        # This class will spawn a thread, so we only want one active at a time.
+        with cls._singleton_lock:
+            if cls._singleton is None:
+                cls._singleton = cls()
+            return cls._singleton
+
+    def __init__(self) -> None:
+        self._active_reads = {}
+        self._thread = threading.Thread(target=self._loop)
+        self._thread.daemon = True
+        self._thread.start()
+
+    def _loop(self):
+        # This loop is scoped to the life of the process, so we rely on process
+        # teardown to pull the rug out from under the daemonic thread running
+        # this function.
+        while True:
+            time.sleep(self._loop_cadence)
+            now = time.time()
+            with self._loop_lock:
+                for w_fd, (timeout, start_time) in tuple(self._active_reads.items()):
+                    if now - start_time >= timeout and w_fd in self._active_reads:
+                        os.write(w_fd, _TIMEOUT)
+                        self.pop(w_fd)
+
+    def pop(self, w_fd: int) -> None:
+        self._active_reads.pop(w_fd, None)
+
+    @classmethod
+    @contextlib.contextmanager
+    def maybe_timeout_read(cls, pipe: "Pipe") -> None:
+        timeout = pipe.timeout
+
+        # Workers should never set a timeout, so in that case we want to exit
+        # without calling `cls.singleton()` so we don't spawn an unnecessary
+        # loop thread.
+        if timeout is None:
+            yield
+            return
+
+        w_fd = pipe.write_fd
+        assert w_fd is not None, "Cannot timeout without write file descriptor."
+        singleton = cls.singleton()
+        with singleton._loop_lock:
+            # This will only occur in the case of concurrent reads on different
+            # threads (not supported) or a leaked case.
+            assert w_fd not in singleton._active_reads, f"{w_fd} is already being watched."
+            singleton._active_reads[w_fd] = (timeout, time.time())
+
+        try:
+            yield
+
+        finally:
+            singleton.pop(w_fd)
+
+
+class Pipe:
```

Comment:
I suspect much of this could be simplified using some combination of select / poll / peek:

https://man7.org/linux/man-pages/man2/poll.2.html

Such as:
https://docs.python.org/3/library/select.html

Reply:
This is assuming your main concern is blocking on an empty pipe.

Reply:
Remember it has to be cross platform, and Windows has a pretty different I/O model. (In particular, peeks / non-blocking reads aren't really a thing.) And I'd rather not have a polling loop going if 99.9% of the time I can rely on OS interrupts.

I do share your concern with the Pipe machinery in general. I'd love to brainstorm if there's a simpler and better way to do it.

Reply:
After offline discussion, we decided to keep this. (And if we think of a better way later it's not hard to swap.)